Include Python 3 bindings into the cmake build and make packages for them

BlaXpirit opened the pull request:
## cmake

#### FindPython3Interp and FindPython3Libs modules have been added
Multiple alternatives have been considered on how to include multiple versions of Python into the build process.

- The first thing that comes to mind is to reuse the standard module *FindPythonInterp* twice with different inputs, however CMake makes this very difficult because this module saves globally cached state and still gives its initial outputs even when ran with different arguments. An effort was made to clear specific cache variables in the middle of the build process to be able to reuse the module, and it seemed to mostly work, even if it was a mess, but in the end it was discovered that Ubuntu 14.04 ships an old version of CMake where *FindPythonLibs* simply does not acknowledge Python 3 yet, even by itself.

The two other alternatives that were dismissed, for hopefully obvious reasons:
- Drop *FindPythonXXX* modules and run hardcoded Python binaries
- Run the whole configure+build process twice with a different flag for the Python version

So, the final decision was to copy the latest versions of the official CMake modules *FindPythonInterp*, *FindPythonLibs* and replace every variable name inside them from `Python`, `PYTHON_` etc. to `Python3`, `PYTHON3_` etc. So even though these modules have global state, the state is now separate for Python 2 and 3.

#### Python bindings are built for Python 2 and 3
This is achieved by looping over two values of a global variable `${PYTHON_VERSION}` &ndash; `""` and `"3"`. *src/pybind/CMakeLists.txt* introduces this variable, calls `find_package(Python${PYTHON_VERSION}Interp ...` and includes the subdirectories, which use the *Distutils* module, which is now aware of this `PYTHON_VERSION` variable.

#### FindCython module has been dropped
The only effect this module did was to fail when a `cython` executable was not found. However, running Cython itself was done through `python setup.py` in a different module, which is a bit of an inconsistency and these two things succeeding are not necessarily correlated. The situation is made worse by the fact that now we need Cython specifically for Python 2 and separately for Python 3.

I acknowledge that there needs to be a way to fail if Cython is not present &ndash; during the configuration phase rather than during the build. Looking for feedback on alternatives. The simple way to do the check would be to run `${PYTHON_EXECUTABLE} -m cython`

#### Summary

This change makes the CMake-based build additionally produce files like *build/lib/cython_modules/lib.linux-x86_64-3.4/`{`rados,cephfs,rbd`}`.cpython-34m.so*

The weakest point of this change is about *FindCython* (discussed above). Other than that, it has been tested on Debian and produces working libraries for both Python versions.


## debian

This is quite straightforward, three new packages have been created: *python3-`{`rados,cephfs,rbd`}`* (based on existing ones). Because Python libraries are installed into a path like */usr/lib/python3.4/dist-packages/*, these *install* files utilize this like so: `usr/lib/python3*/dist-packages/rados.so`. Additionally, the file lists for Python 2 needed to be changed to include a *2* so they don't pick up files for both versions of Python.
The newly added packages get a line [`Depends: ${python3:Depends}`](https://wiki.debian.org/Python/LibraryStyleGuide#Python_versions)

Some build dependencies were added: [*python3-dev*](https://packages.debian.org/search?keywords=python3-dev), [*python3-setuptools*](https://packages.debian.org/search?keywords=python3-setuptools), [*cython3*](https://packages.debian.org/search?keywords=cython3).

This change has **not** been confirmed working, but to the best of my knowledge there should be no problems.


## rpm

Three new packages have been created: *python34-`{`rados,cephfs,rbd`}`* (based on existing ones). Well, the version *34* is not specified anywhere, but my understanding is that the `%{python3_pkgversion}` macro expands to *34* on CentOS and to *3* on Fedora. `%{python3_sitearch}` and `%{python3_sitelib}` were utilized. [*python3-numpy.spec*](http://pkgs.fedoraproject.org/cgit/rpms/python3-numpy.git/tree/python3-numpy.spec?h=epel7) served as an example of these techniques.

Added dependencies:

- RHEL/CentOS:
[*python34*](https://dl.fedoraproject.org/pub/epel/7/x86_64/p/python34-3.4.3-4.el7.x86_64.rpm)
    - build:
[*python34-devel*](https://dl.fedoraproject.org/pub/epel/7/x86_64/p/python34-devel-3.4.3-4.el7.x86_64.rpm), [*python34-setuptools*](https://dl.fedoraproject.org/pub/epel/7/x86_64/p/python34-setuptools-19.2-3.el7.noarch.rpm), [*python34-Cython*](https://dl.fedoraproject.org/pub/epel/7/x86_64/p/python34-Cython-0.23.5-1.el7.x86_64.rpm)
- Fedora:
[*python3*](https://admin.fedoraproject.org/pkgdb/package/rpms/python3/)
    - build:
[*python3-devel*](https://apps.fedoraproject.org/packages/python3-devel), [*python3-setuptools*](https://apps.fedoraproject.org/packages/python3-setuptools), [*python3-Cython*](https://apps.fedoraproject.org/packages/python3-Cython)
- OpenSUSE:
[*python3*](http://download.opensuse.org/tumbleweed/repo/oss/suse/x86_64/python3-3.5.1-3.4.x86_64.rpm)
    - build:
[*python3-devel*](http://download.opensuse.org/tumbleweed/repo/oss/suse/x86_64/python3-devel-3.5.1-3.3.x86_64.rpm), [*python3-setuptools*](http://download.opensuse.org/tumbleweed/repo/oss/suse/noarch/python3-setuptools-23.1.0-1.1.noarch.rpm), [*python3-cython*](http://download.opensuse.org/tumbleweed/repo/oss/suse/x86_64/python3-Cython-0.24-2.2.x86_64.rpm)

This change has **not** been tested at all.

The big problem with it is that on CentOS all mentioned Python3-related packages come from EPEL (not readily available to install by default). And they are mandatory for the build, so it may become generally more difficult to build packages.

tchaikov commented on src/pybind/CMakeLists.txt:
```diff
@@ -1,33 +1,42 @@
+foreach(python_version "" "3")
+  set(PYTHON_VERSION ${python_version})
+
+  find_package(Python${PYTHON_VERSION}Interp ${PYTHON_VERSION} QUIET)
```
just want to understand why the stock `FindPythonInterp.cmake` does not work in this case. as it lists "3.4 3.3 3.2 3.1 3.0" as supported python3 versions (in cmake 3.3.0). maybe the `FindPythonInterp.cmake` packaged by the supported distros is too old?

tchaikov commented:
> The big problem with it is that on CentOS all mentioned Python3-related packages come from EPEL (not readily available to install by default). And they are mandatory for the build, so it may become generally more difficult to build packages.

@BlaXpirit iiuc, these python3 packages are all required for developer, maintainer or packager. not a requirement for the user if he/she don't want to use the python3 bindings, right?

@kdreyer could this be a blocker? iirc, the lttng-ust-devel and libbabeltrace-devel also live in EPEL repo.

BlaXpirit commented:
> these python3 packages are all required for developer, maintainer or packager. not a requirement for the user if he/she don't want to use the python3 bindings

Correct

jdurgin commented:
Talked a little about the latest build failure on IRC - it's just down to 'unpackaged files' for python3 versions of ceph_argparse, ceph_daemon, and ceph_rest_api. Right now the python2 versions are included in the ceph-common and ceph-mon packages.

The ceph cli does need ceph_daemon and ceph_argparse, so testing it with python3 will require setting PYTHONPATH or making python3 packages for those. Perhaps we could include them in python3-rados since they are wrappers helping expose basic rados functionality.

BlaXpirit commented:
So yes, I have continued work on this, ironing out bugs one by one:

- Incorrect file names to be packaged (`rbd.so` vs `rbd.cpython*.so`)
- Had to hardcode package names in build dependencies for RPM because macros like `python%{python3_pkgversion}` don't work there.
- Bug with virtualenv (merged [here](https://github.com/ceph/ceph/pull/10289/commits/2699b6d89b8c1c193fd86b5233d1ea86458753a0))
- Specified Python 2 explicitly to FindPythonInterp just in case (well it would break on Arch Linux...)

And I've reworked *FindCython.cmake*, now it only checks for existence of a 'cython' module, by running a corresponding Python executable (works for 2 and 3).

The build succeeded on Ubuntu, but seems like it's run into yet another hurdle on CentOS, as described by *jdurgin*.

jdurgin commented:
There are some related changes in https://github.com/ceph/ceph/pull/10322

particularly ```debian: remove ${shlibs:Depends} from python-* packages``` should be done to these new packages too

jdurgin commented:
needs a rebase

BlaXpirit commented:
I don't think this can be merged anyway because CentOS packages wouldn't build (well, without the other pull request everything would fail). If this were to be merged, I'd need to add some temporary workarounds from my test branch.

BlaXpirit commented:
We discussed this with @jdurgin, and I resolved the mentioned problems. The remaining 2 files went into a `python3-ceph-argparse` package (also added it for debian)

This is now ready for merging. Debian and RPM packages build successfully.

jdurgin commented:
jenkins: retest this please

jdurgin commented on src/rocksdb:
```diff
@@ -1 +1 @@
-Subproject commit 6797e6ffacd90470140f64cea068970bb3a30aaf
+Subproject commit 5647fa427c4eab87c0caf8bcbb91f5846a4623f9
```
accidentally got a submodule update in there

BlaXpirit replied:
Hmm, these are nasty. I actually noticed it and it seemed like I was able to get rid of it, but no, it's still there.

jdurgin commented:
looks good to me with the submodule update removed. @smithfarm @ktdreyer do these packaging changes look ok to you?

tchaikov commented on src/pybind/CMakeLists.txt:
```diff
@@ -1,33 +1,51 @@
 
-add_custom_target(cython_modules ALL
-  DEPENDS cython_rados cython_cephfs cython_rbd)
+  find_package(Python${PYTHON_VERSION}Interp ${python_version} QUIET)
```
why not `REQUIRED`?

BlaXpirit replied:
Just copied what was there before - no `REQUIRED` but a custom error message right after. It might make it clearer what version of Python is needed.
If you think that needs to be changed, please reply again.

tchaikov replied:
the default message emitted by `find_package_handle_standard_args()` would suffice.

tchaikov replied:
@BlaXpirit this is not addressed.

tchaikov commented on debian/control:
```diff
@@ -57,6 +58,8 @@ Build-Depends: btrfs-tools,
                python-nose,
 	       python-setuptools,
                python-sphinx,
+               python3-dev,
```
better off using `python3-all-dev`. see http://debian-python.readthedocs.io/en/latest/debian-policy.html#build-dependencies .

BlaXpirit replied:
OK, will be changed

tchaikov commented on debian/control:
```diff
@@ -654,6 +669,18 @@ Description: Python libraries for the Ceph librbd library
+ storage system that runs on commodity hardware and delivers object,
+ block and file system storage.
+ .
+ This package contains Python libraries for interacting with Ceph's
```
might want to note that

> This package contains the Python 3 library for interacting with Ceph's RBD block device library.

tchaikov replied:
and i'd also recommend change the description of `python-rbd` to note that it contains the Python 2 library for interacting with Ceph's RBD block device library.

the same applies to other python libraries.

BlaXpirit replied:
OK

tchaikov commented on ceph.spec.in:
```diff
@@ -1363,6 +1427,22 @@ ln -sf %{_libdir}/librbd.so.1 /usr/lib64/qemu/librbd.so.1
+%files -n python%{python3_pkgversion}-ceph-argparse
+%defattr(-,root,root,-)
+%{python3_sitelib}/ceph_argparse.py
+%{python3_sitelib}/__pycache__/ceph_argparse.cpython*.py*
```
how does ceph_argparse.py interact with cython? and does `install()` command of cmake creates the byte compiled files in `__pycache__`?

BlaXpirit replied:
*ceph_argparse.py* does not interact with Cython (not to be confused with *cpython*).

RPM build process creates the byte compiled files in *\__pycache__*.

tchaikov replied:
oh, i misread cpython as cython. thanks for the explanation!

BlaXpirit commented:
I addressed the comments.

Need to rebase again, but I have a comment about [one of the new commits](https://github.com/ceph/ceph/commit/bb163e98#commitcomment-18586068)

smithfarm commented on ceph.spec.in:
```diff
@@ -405,7 +417,17 @@ License:	LGPL-2.0
+This package contains Python 2 libraries for interacting with Cephs RADOS
+object store.
+
+%package -n python%{python3_pkgversion}-rados
```
I'm concerned that this `%python3_pkgversion` macro might not be defined. Could you add something like

```
%{!?python3_pkgversion: %global python3_pkgversion 3}
```

near the top, just above (or below) the similar lines that ensure `%_udevrulesdir` and `%tmpfiles_create` have sane values?

BlaXpirit replied:
Done

smithfarm commented on ceph.spec.in:
```diff
@@ -153,6 +156,15 @@ BuildRequires:  openldap-devel
+BuildRequires:	python3-setuptools
+BuildRequires:	python3-Cython
+%endif
 %endif
```
The BuildRequires for suse, fedora, and centos are all the same, I gather. In that case, instead of the above modifications to the suse and fedora/rhel sections, which introduce duplication, could you just put the following lines here:

```

# python34-... for rhel, python3-... for all other supported distros
%if 0%{?rhel}
BuildRequires:	python34-devel
BuildRequires:	python34-setuptools
BuildRequires:	python34-Cython
%else
BuildRequires:	python3-devel
BuildRequires:	python3-setuptools
BuildRequires:	python3-Cython
%endif
```

BlaXpirit replied:
I don't understand what exactly you want added. Just the comment?

smithfarm replied:
yes, after moving the conditional block and deleting the duplicate conditionals from the suse conditional block, the comment is only thing that is new.

smithfarm replied:
I'm trying to eliminate the duplication of these three lines:

```
BuildRequires:  python3-devel
BuildRequires:  python3-setuptools
BuildRequires:  python3-Cython
```

smithfarm commented on ceph.spec.in:
```diff
@@ -466,7 +488,17 @@ Requires:	librbd1 = %{epoch}:%{version}-%{release}
 Obsoletes:	python-ceph < %{epoch}:%{version}-%{release}
 %description -n python-rbd
-This package contains Python libraries for interacting with Cephs RADOS
+This package contains Python 2 libraries for interacting with Cephs RADOS
```
Please propagate this change `s/Python libraries/Python 2 libraries/g` everywhere appropriate in both rpm and deb.

BlaXpirit replied:
I specifically made these changes only in the description and not the summary. So it looks like you think the summaries also all should be changed?

smithfarm replied:
Yes. There are contexts where the summary is the only thing the user sees. It seems to me that adding the " 2" or " 3" to the summary only increases the length by two characters, yet adds significant information value. It's just an idea - I don't feel strongly about it.

BlaXpirit replied:
Done

smithfarm commented on ceph.spec.in:
```diff
@@ -504,9 +536,29 @@ Requires:	libcephfs1 = %{epoch}:%{version}-%{release}
 file system.
 
+%package -n python%{python3_pkgversion}-cephfs
+Summary:	Python libraries for Ceph distributed file system
```
Python 2 libraries

(lots of these are left - you can find them by grepping for "Python libraries")

smithfarm commented on ceph.spec.in:
```diff
@@ -504,9 +536,29 @@ Requires:	libcephfs1 = %{epoch}:%{version}-%{release}
+file system.
+
+%package -n python%{python3_pkgversion}-ceph-argparse
+Summary:    Python types and routines used by the Ceph CLI
```
Python 3 types . . .

BlaXpirit replied:
I was going for specifying the version only in the description.

smithfarm replied:
Why?

smithfarm commented on ceph.spec.in:
```diff
@@ -142,6 +142,9 @@ BuildRequires:  libopenssl-devel
 BuildRequires:	python-Cython
+BuildRequires:	python3-devel
+BuildRequires:	python3-setuptools
+BuildRequires:	python3-Cython
```
delete this

smithfarm commented on ceph.spec.in:
```diff
@@ -153,6 +156,15 @@ BuildRequires:  openldap-devel
+BuildRequires:	python3-devel
+BuildRequires:	python3-setuptools
+BuildRequires:	python3-Cython
+%endif
```
move this conditional block (`%if 0%{?rhel}`) outside of the outer conditional block

BlaXpirit replied:
Done

smithfarm commented on ceph.spec.in:
```diff
@@ -153,6 +156,15 @@ BuildRequires:  openldap-devel
 BuildRequires:  openssl-devel
 BuildRequires:  redhat-lsb-core
 BuildRequires:	Cython
```
does Fedora/RH/CentOS have a "Cython" package in addition to "python-Cython" ?

smithfarm replied:
and is this "Cython" package really needed?

BlaXpirit replied:
OpenSUSE has *python-Cython*, Fedora/RH/CentOS have *Cython*. I think all is correct here.

Yes, it is still needed to build Python 2 extension libraries.

smithfarm replied:
Cython for Python 2, and python3-Cython for Python 3 - got it.

BlaXpirit commented:
Thanks for the comments. I think I addressed everything so far.

smithfarm commented on ceph.spec.in:
```diff
@@ -1266,6 +1317,12 @@ fi
+%{python3_sitearch}/rados.cpython*.so
+%{python3_sitearch}/rados-*.egg-info
+
+#################################################################################
```
the `#######` divider lines are being dropped in #9757, so can you remove them from your patch (just the new ones that you're adding)

BlaXpirit replied:
I don't see how to do this consistently, because the new packages go in between the old packages.

I'm afraid these two PRs will conflict no matter what, so I think it makes more sense to keep things as is.

smithfarm replied:
You're right - I'll wait for this PR to be merged, then update #9757.

BlaXpirit commented:
Welp, something went horribly wrong. http://gitbuilder.sepia.ceph.com/gitbuilder-ceph-rpm-centos7-amd64-basic/log.cgi?log=09f27ab6b8c546b8c2e48fbc90dee5cb45f65235

smithfarm commented:
@BlaXpirit Debian doesn't care what junk you leave lying around, but RPM has a "Fascist build policy" [1] which means all of those "installed but unpackaged files" have to dealt with.

[1] https://docs.fedoraproject.org/ro/Fedora_Draft_Documentation/0.1/html/RPM_Guide/ch09s05s07.html

BlaXpirit commented:
@smithfarm I'm well aware of that, and I've had successful builds previously. The problem is now for some reason Python 2 libraries end up in Python 3 folders and vice versa.

BlaXpirit commented:
@tchaikov I'm almost sure the problems are caused by a conflict with https://github.com/ceph/ceph/pull/10568 and I don't understand what you did there.

tchaikov commented:
> I'm almost sure the problems are caused by a conflict with #10568

@BlaXpirit so do i.

> I don't understand what you did there.

as the commit message put, i don't want to repeat the logic to build the path (`"lib.{platform}-{version[0]}.{version[1]}"`) everywhere.

tchaikov commented:
fix pushed to ceph.git/wip-cmake.

jdurgin commented:
@tchaikov seems like we'd need to add the python version to the name of that subdir again for this PR to work

BlaXpirit commented:
Yeah, @tchaikov, if I had to guess, I would say this wouldn't solve the problem.

Anyhow, just to confirm, with that commit [reverted](https://github.com/ceph/ceph/commits/58906461047baa2c79279eaa454d7cc8ec4be49f) it [builds fine](http://gitbuilder.sepia.ceph.com/gitbuilder-ceph-rpm-centos7-amd64-basic/log.cgi?log=58906461047baa2c79279eaa454d7cc8ec4be49f)

tchaikov commented:
> seems like we'd need to add the python version to the name of that subdir again for this PR to work

@jdurgin yes, the cythonized modules for python3 would be overwritten by the ones for python2 or otherwise. seems i am sort of reinventing the wheel if i put the python version back into the path. but right, we should.

> if I had to guess, I would say this wouldn't solve the problem.

@BlaXpirit sorry, i don't follow you. i agree with josh, that to make your PR work, we need encode the python version into the path. but why it does not solve "the problem"?

jdurgin commented:
@tchaikov another idea is calling whatever distutils method generates that path to replace the hardcoding in AddCephTest and the duplicate logic in ceph.in. Is that feasible?

tchaikov commented:
@jdurgin, i investigated that approach before: the "distutils method" is not a standalone method and hence not callable.

tchaikov commented:
```py
    def finalize_options(self):
        if self.plat_name is None:
            self.plat_name = get_platform()
        else:
            # plat-name only supported for windows (other platforms are
            # supported via ./configure flags, if at all).  Avoid misleading
            # other platforms.
            if os.name != 'nt':
                raise DistutilsOptionError(
                            "--plat-name only supported on Windows (try "
                            "using './configure --help' on your platform)")

        plat_specifier = ".%s-%s" % (self.plat_name, sys.version[0:3])
```

quote from `distutils/command/build.py`

BlaXpirit commented:
Alright, thanks, @tchaikov, looks like this is ready to go now. Gitbuilders succeeded.

smithfarm commented:
lgtm `Reviewed-by: Nathan Cutler <ncutler@suse.com>`

BlaXpirit commented:
No, the Jenkins build indicates a problem now.

jdurgin commented:
@BlaXpirit if you rebase on master it should work - looks like you don't have @tchaikov's last fix in your branch

BlaXpirit commented:
@jdurgin, but I do have that fix in my branch

BlaXpirit commented:
Alright, I figured it out ↑

Now Jenkins and Gitbuilders ([wip-cmake-py3](https://github.com/ceph/ceph/compare/wip-cmake-py3)) are happy

tchaikov commented on cmake/modules/FindPython3Interp.cmake:
```diff
@@ -0,0 +1,148 @@
+#.rst:
```
where is this module used?

BlaXpirit replied:
https://github.com/ceph/ceph/pull/10208/files#diff-95e351a3805a1dafa85bf20b81d086e6R283

tchaikov replied:
oh, okay.

tchaikov commented:
lgtm.

jdurgin commented:
nice work! lgtm too. I'll just run this through some of the pybind tests from a few suites
